### PR TITLE
Query: Adds type-markers with count and length for large arrays

### PR DESCRIPTION
**Pull request by Maya-Painter**
## Description

Updating JsonBinaryWriter to match query backend behavior. For large arrays, type-markers that encode both the array length and item count are used.

Large arrays are defined as:
- Arrays of 1-byte length and of more than 16 items
- Arrays of 2-byte length and of more than 16 items, or if the length exceeds 4 KB
- All arrays of 4-byte length

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Json/JsonWriter.JsonBinaryWriter.cs [hunk 1]
```diff
@@ -248,7 +248,7 @@ private enum RawValueType : byte
             /// This flag just determines whether you want to serialize the count, since it's optional and up to the user to make the
             /// tradeoff between O(1) .Count() operation as the cost of additional storage.
             /// </summary>
-            private readonly bool serializeCount;
+            private readonly bool serializeCountOverride;
```
Comment (Maya-Painter) [inline Microsoft.Azure.Cosmos/src/Json/JsonWriter.JsonBinaryWriter.cs:251]: Is this something we want to keep?  As far as I can tell, all the callers set this to false.

**Comment by Maya-Painter** [comment]
@microsoft-github-policy-service agree company="Microsoft"

File: Microsoft.Azure.Cosmos/src/Json/JsonWriter.JsonBinaryWriter.cs [hunk 2]
```diff
@@ -611,16 +611,25 @@ private void WriteArrayOrObjectEnd(bool isArray)
...
+
+                        // 1 byte length
                         int bytesToWrite = JsonBinaryEncoding.TypeMarkerLength
                             + JsonBinaryEncoding.OneByteLength
-                            + (this.serializeCount ? JsonBinaryEncoding.OneByteCount : 0);
+                            + (serializeCount ? JsonBinaryEncoding.OneByteCount : 0);
+
+                        // Move the buffer forward if count is serialized
+                        if (serializeCount)
+                        {
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Json/JsonWriter.JsonBinaryWriter.cs:623]: please remove this condition

**Comment by sboshra** [comment]
>         [Owner("brchon")]

please move to your name

---
Refers to: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Json/JsonWriterTests.cs:1649 in 75de8e7. [](commit_id = 75de8e73dc7d771a48b500b1ac80d63524375e6e, deletion_comment = False)

**Review by sboshra — APPROVED** [review]
:shipit:

**Pull request merged**